confgenerator/otel/modular.go

```diff
@@ -97,7 +98,12 @@ func configToYaml(config interface{}) ([]byte, error) {
 	if err := mapstructure.Decode(config, &outMap); err != nil {
 		return nil, err
 	}
-	return yaml.Marshal(outMap)
+	return yaml.MarshalWithOptions(
+		outMap,
+		yaml.CustomMarshaler[commonconfig.Secret](func(s commonconfig.Secret) ([]byte, error) {
```

Comment:
I just realized that [`modular.go:97`](https://github.com/GoogleCloudPlatform/ops-agent/blob/2b7466370c9aad268489f3ce4e1b43bdc218aef1/confgenerator/otel/modular.go#L97) may be addressing the same problem a different way. Should we be using a uniform approach? Want to at least add a TODO?

Reply:
I think you're right; the output is identical if I remove the mapstructure decode first. So I think I'll remove that here and update the comment.

Reply:
Actually, maybe I should ask @ridwanmsharif what that was here for and if there's a reason for it I'm not seeing.

Reply:
Is the output also identical if you remove the custom marshaller here (and only use `mapstructure.Decode`)? I suspect the main issue was that `deepCopy` marshalled the unmodified config struct…

Reply:
In that scenario (no custom unmarshaler, just mapstructure decode), the output still has `secret`.

Reply:
Ah, ok. We can clean this up another time.

Reply:
I'm not aware of any reason why that was there. I suspect it has something to do with us using OTel structs directly in the receiver configs (like https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/b39f83d56853ca03fc8edd196eedf65e5046b9fb/receiver/apachereceiver/config.go#L16-L20) and if they use `mapstructure` annotations, they would need to be decoded appropriately using the annotations before we marshal them into yaml.

With that said, there is nothing prometheus specific about this, and wasn't there because of the prometheus structs. I suspect if none of our golden files change when we remove this `mapstructure.Decode` that might just be because the structs in the pipeline don't have any complicated decoding rules defined. But that is by convention and keeping things as is will keep this less error prone.

Reply:
Thanks Ridwan, sounds like we should leave it in there for now.